### ISSUE TITLE
new Map(Iterable[K, V]) with complex value type V can have arbitrary extra properties on each V

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/TypeScript#49500]
# Bug Report

### 🔎 Search Terms

map from iterable 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about "map". The issue exhibits on the nightly playground.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220611&ssl=9&ssc=55&pln=8&pc=1#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyARnFAFzIDOYUoA5gDQFGkBelIArgLbHQEAvgQIB6UchBZoUdFEbIIANwghkYABbou9DcgCOXAB7JgVSejDI4yJXAA2wACbIADrNfQwAT2ToYaJgE9hBWMJgAsnCulFGuADw0dCBMgegAfMgAvJIQAO7IcQAUANolAETh6OUK2KQUyOX1NSRwHMgAjAqGRpTlPeWCALpDAJQA3GISmihUXMS0iGDAKsgIcFQoZmtwOhpgBFVx2bkFxWWVmC11ZH3N3cZ9A8Njk-jipgEzppBQcMQhUzmVxcewhFzaKxwEAuHyeJwKb4yORAnZ7MAAOjW2nsTim1nsVCwiA0wGUKE02zywE01moXBgMGAJn8yAABhtkCUkgwFBh0EMSkM2dZiOgVMFQsgqgBJX7-EKUbm0XlpQVDE4XKrXep3MgtdiULoGR6NZ4jfGbMDLFLIam00B2RwuJ1cCDmbYOIlo3QHfBHaIYq1FJr62q6xr3VrtY09J7GQajfE0gDk5jyclpUht9HUGjgVncwB4NJWKFdFO8nnMZBQCA0EAQAGsIC5ORAjJ4EJA8fgQlYeSk4rFookVSkFIP6JkciB8oVoqUKtqFB0Rkn8EA)

### 💻 Code

```ts
interface Foo {
  bar: string,
  baz: number
}

// no error, even though qux is not a valid property of Foo
let fooMap: Map<string, Foo> = new Map([["foo", {bar: "bar", baz: 1, qux: "qux"}]]);
// the subtractive case is caught
fooMap = new Map([["foo", {bar: "bar", qux: "qux"}]]);
// if the iterable is pulled out and typed, the error is caught. could
// also achieve this with a suffix of `as [string, Foo][]` above
let fooIterable: [string, Foo][] = [["foo", {bar: "bar", baz: 1, qux: "qux"}]]
// setting with invalid values is also caught
fooMap.set("bar", {bar: "bar", baz: 1, qux: "qux"})
// it's worth noting that primitive value types are checked as expected
let stringMap: Map<string, string> = new Map([["foo", 1]])
```

### 🙁 Actual behavior

`identifier: Map<K, V> = new Map(Iterable[K, V])` allows each `V` in the iterable to have arbitrary extra properties.

### 🙂 Expected behavior

`identifier: Map<K, V> = new Map(Iterable[K, V])` should check that each value in the iterable is a valid `V`.

**Comment by MartinJohns** [comment]
Objects are not sealed. They're allowed to have arbitrary extra properties. This is unrelated to `Map<>`.

You want #12936, but don't expect it to happen any time soon.

**Comment by thisisrandy** [comment]
Right, I think I've understood. Basically, it boils down to I can do this
```ts
interface Foo {
  bar: string,
  baz: number
}

const fooLike = { bar: "bar", baz: 1, qux: 1 };
const foo: Foo = fooLike;
```
But not this:
```ts
const badFooLiteral: Foo = { bar: "bar", baz: 1, qux: 1 };
```
The `Object literal may only specify known properties...` error was pushing me in the wrong direction on a still fuzzy (but now quite clear) point about structural typing.

If I may, do the docs cover this well? I would expect to see something highlighting the object literal case e.g. on the [Type Compatiblity](https://www.typescriptlang.org/docs/handbook/type-compatibility.html) page, but I only see two step examples.

**Comment by MartinJohns** [comment]
> I think I've understood. Basically, it boils down to I can do this

Yep, exactly.

> The Object literal may only specify known properties... error was pushing me in the wrong direction on a still fuzzy (but now quite clear) point about structural typing.

It very often leads to this kind of misunderstanding. It should be considered more of a linter feature, not a type-safety feature.

> If I may, do the docs cover this well? I would expect to see something highlighting the object literal case e.g. on the [Type Compatiblity](https://www.typescriptlang.org/docs/handbook/type-compatibility.html) page, but I only see two step examples.

It's shown under "Starting out", the basic rule is mentioned there.  
Otherwise:
![image](https://user-images.githubusercontent.com/5269069/173253392-cd23aeb2-1a84-46e7-a889-7927ed3d58c7.png)

**Comment by thisisrandy** [comment]
Great, thanks for your help. I see some stuff on e.g. [TypeScript for JS Programmers](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html) that deals with *misspelling* of properties, but I'm not sure that properly nails home the point about *extra* properties when the required ones are already properly specified. I'll think on where a note about this might fit on the Type Compatibility page and submit a PR.

**Comment by MartinJohns** [comment]
Your example with "Foo" is literally the the second example on that page: https://www.typescriptlang.org/docs/handbook/type-compatibility.html#starting-out

```typescript
interface Foo {
  bar: string,
  baz: number
}

const fooLike = { bar: "bar", baz: 1, qux: 1 };
const foo: Foo = fooLike;
```
---
```typescript
interface Pet {
  name: string;
}
let pet: Pet;
// dog's inferred type is { name: string; owner: string; }
let dog = { name: "Lassie", owner: "Rudd Weatherwax" };
pet = dog;
```

**Comment by thisisrandy** [comment]
No, I see that. My point was that 
```ts
// Object literal may only specify known properties, and 'qux' does not exist in type 'Foo'
const badFooLiteral: Foo = { bar: "bar", baz: 1, qux: 1 };
```
was *not* highlighted. The example
```ts
interface User {
  name: string;
  id: number;
}
 
const user: User = {
  username: "Hayes",
  // Type '{ username: string; id: number; }' is not assignable to type 'User'.
  // Object literal may only specify known properties, and 'username' does not exist in type 'User'.
  id: 0,
};
```
[here](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html) is basically the same thing, but that may not be obvious to new users learning the core concepts, since a *misspelled* property doesn't necessarily feel the same as an *extra* property.

Here's what happened to me:
1. I thought I understood that an object is assignable to a type as long as the object is a weak structural superset of the type (or `any`).
2. The `Object literal may only specify known properties...` error made me second-guess myself, because naively, it seemed the same as (1).
3. I was unable to understand why `let fooMap: Map<string, Foo> = new Map([["foo", {bar: "bar", baz: 1, qux: "qux"}]]);` was fine because the logic of assignability had gotten fuzzed.

Hence, I want to highlight the object literal assignment case somewhere clearly in the docs. If that's already done and I somehow missed it, please point it out, but the example you highlighted was absorbed on reading on didn't help to keep me from confusing myself here.
